File: 백준/Silver/1074. Z/Z.py
import sys

input = sys.stdin.readline
n, r, c = map(int, input().split())

def f(sr, sc, er, ec):
  global r, c
  total_count = (er - sr + 1) * (ec - sc + 1)
  each_block_count = total_count // 4

  midr = (sr + er) // 2
  midc = (sc + ec) // 2

  # r, c가 영역 안에 있어야만 루프를 돌 필요가 있음
  if not (sr <= r <= er and sc <= c <= ec):
    return 10e9

  if (sr == er and ec == sc):
    return 1

  return min(
    f(sr, sc, midr, midc), 
    each_block_count + f(sr, midc + 1, midr, ec), 
    each_block_count * 2 + f(midr + 1, sc, er, midc), 
    each_block_count * 3 + f(midr + 1, midc + 1, er, ec)
  )


print(f(0, 0, 2**n-1, 2**n-1) - 1)

File: 백준/Silver/1074. Z/Z.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split(' ')
  .map(Number);

const [N, r, c] = input;
const size = 2 ** N;

function solution(startR, startC, endR, endC) {
  const quadrantCount = ((endR - startR + 1) * (endC - startC + 1)) / 4;
  const midR = parseInt((startR + endR) / 2);
  const midC = parseInt((startC + endC) / 2);

  if (r < startR || r > endR || c < startC || c > endC) return Infinity;

  if (startR == endR && startC == endC) return 1;

  return Math.min(
    solution(startR, startC, midR, midC),
    quadrantCount + solution(startR, midC + 1, midR, endC),
    quadrantCount * 2 + solution(midR + 1, startC, endR, midC),
    quadrantCount * 3 + solution(midR + 1, midC + 1, endR, endC)
  );
}

console.log(solution(0, 0, size - 1, size - 1) - 1);
